### PR TITLE
Caffe to MXNet code translator

### PR DESCRIPTION
**Pull request by indhub**
## Description ##
CaffeTranslator is a tool to translate Caffe code (training prototxt) to MXNet python code. 

More info here: https://github.com/indhub/mxnet/blob/caffe_translator/tools/caffe_translator/README.md

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [x] For user-facing API changes, API doc string has been updated. For new C++ functions in header files, their functionalities and arguments are well-documented. 
- [x] To my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by indhub** [comment]
@pracheer @Roshrini @sandeep-krishnamurthy @nswamy @madjam

**Comment by piiswrong** [comment]
Why does this require Java? That's a huge additional dependency. Most researchers don't have it.

**Comment by indhub** [comment]
This tool uses Java because libraries like [ANTLR](http://www.antlr.org/) and [StringTemplate](http://www.stringtemplate.org/) make it easy to write source code translator.

Java is usually easy to install. For example, on Ubuntu:
```
sudo apt-get install default-jre
```

**Comment by pracheer** [comment]
Overall looks good to me.

Few comments:
- Some files don't seem to have any license. E.g. [batchnorm.st](https://github.com/indhub/mxnet/blob/caffe_translator/tools/caffe_translator/src/main/resources/templates/batchnorm.st).
- Some sort of guidance/document on how the translator internally works will be useful since from what I see there are bunch of components for this translator. Such details will help 1. reviewers better do the review and 2. for anyone else who wants to fix any issues with the translator later on when you are not around.
- No unit tests?
- Do you think it'll be good to exit in case of exceptions like [this](https://github.com/indhub/mxnet/blob/caffe_translator/tools/caffe_translator/src/main/java/io/mxnet/caffetranslator/Solver.java#L56) or [this](https://github.com/indhub/mxnet/blob/caffe_translator/tools/caffe_translator/src/main/java/io/mxnet/caffetranslator/Converter.java#L284)? Basically fail fast in case of issues? Or would it make more sense to generate a code-comment which says "// Couldn't convert <blah> lr policy" and continue. Also, with that in mind, what do you think about having a force flag to the translator which if present ignores such errors and continues instead of exiting?
- Probably integrate FAQ with the main README since main README doesn't have too much info? (this may just be my personal preference though).
- Out of curiosity what happens when [this](https://github.com/indhub/mxnet/blob/caffe_translator/tools/caffe_translator/src/main/java/io/mxnet/caffetranslator/Converter.java#L362) returns null?

Few points mainly around style:
- Do you think having better class docs would be helpful for anyone trying to maintain the code?
- Generally good to have braces for if statements even if it contains only one statement. Also spaces b/w if and () would be good everywhere. See [google braces guide](https://google.github.io/styleguide/javaguide.html#s4.1.1-braces-always-used) and [whitespace guide](https://google.github.io/styleguide/javaguide.html#s4.6.2-horizontal-whitespace) for more. I'm assuming fixing now would require too much effort w/o giving enough in return so feel free to choose accordingly.
- I'm seeing some inconsistencies regarding use of final modifier. E.g. [here](https://github.com/indhub/mxnet/blob/caffe_translator/tools/caffe_translator/src/main/java/io/mxnet/caffetranslator/Converter.java#L48) and [here](https://github.com/indhub/mxnet/blob/caffe_translator/tools/caffe_translator/src/main/java/io/mxnet/caffetranslator/CreateModelListener.java#L40). Generally good to be strict unless we there is a reason to not be.
- (Minor) Instead of calling this [GenHelper](https://github.com/indhub/mxnet/blob/caffe_translator/tools/caffe_translator/src/main/java/io/mxnet/caffetranslator/GenHelper.java#L33) GeneratorHelper may be a more appropriate name since it clearly says what is it about.
- Also it would be good to expand the imports such as [this](https://github.com/indhub/mxnet/blob/caffe_translator/tools/caffe_translator/src/main/java/io/mxnet/caffetranslator/GenHelper.java#L33).

Feel free to ignore comments in the interest of time/effort if necessary or if you don't agree with them :)

**Comment by piiswrong** [comment]
Python has some pretty good string libraries. Since you are generating python code, would it make more sense to do it in python? It also avoid the extra jvm dependency, which makes it much harder to use for windows/centos users etc

**Comment by indhub** [comment]
> Some files don't seem to have any license. E.g. batchnorm.st.

These files are string templates. If I place license in these files, the license will go into the generated code. 

> Some sort of guidance/document on how the translator internally works will be useful since from what I see there are bunch of components for this translator. Such details will help 1. reviewers better do the review and 2. for anyone else who wants to fix any issues with the translator later on when you are not around.

> Do you think having better class docs would be helpful for anyone trying to maintain the code?


Totally agree. These are on my list but haven't had the time to do it. Will get to it after the first release. 

> No unit tests?

Unit tests unfortunately is tricky to write. For example, to test the convolution layer generator, I could do some string processing on the output to validate it but that will be too fragile (will break when output changes a little). Ideally, test code should use MXNet to load the generated symbol and verify it but loading incomplete networks (like just one layer) might not be possible. Testing the generated code using a parser is a chicken-egg problem. 

That said, I do have tests at macro level - tests that will translate a full caffe network, run training using the generated code and test for convergence. Those tests are available [here](https://github.com/indhub/caffe_translator_tests). I haven't created a PR for those yet because I need to figure out some licensing related issues. The test code translates some Caffe example code taken from Caffe repo and tests for convergence. I need to figure out what I need to do to use caffe example code for testing without infringing on the license. Planning to merge these tests after the first release. 

> Do you think it'll be good to exit in case of exceptions like this or this? Basically fail fast in case of issues? Or would it make more sense to generate a code-comment which says "// Couldn't convert lr policy" and continue. Also, with that in mind, what do you think about having a force flag to the translator which if present ignores such errors and continues instead of exiting?

In the first case, it will be nice to print a error message saying the caffe prototxt is not well formated. The translator currently assumes the Caffe code is well formatted and it is runnable using Caffe. 

In the second case, it will be nice to add a comment to the generated code indicating the problem. 

You are right.  Adding a comment in the generated code or printing to the console and proceeding with the translation is more preferable than aborting. This is because, as a user, the generated code is helpful to me even if certain parts is not translated. I can manually add few lines to complete the translation. 

These changes are now done. 

> Probably integrate FAQ with the main README since main README doesn't have too much info? (this may just be my personal preference though).

I'll move the list of supported layers to README. Will leave the question about why caffe in the FAQ page. I don't think that is important enough for the README page. I do get your point about there being less content in faq. But more might get added by the time this work is complete.

> Out of curiosity what happens when this returns null?

If there is no data layer in the Caffe prototxt (maybe because the prototxt uses ImageData), this could return null. In that case, the generated code will not have a data iterator. Supporting more data layers like ImageData is in the roadmap but not for the first release.

> Generally good to have braces for if statements even if it contains only one statement. Also spaces b/w if and () would be good everywhere. See google braces guide and whitespace guide for more. I'm assuming fixing now would require too much effort w/o giving enough in return so feel free to choose accordingly.

Fixed.

> I'm seeing some inconsistencies regarding use of final modifier. E.g. here and here. Generally good to be strict unless we there is a reason to not be.

Fixed.

> (Minor) Instead of calling this GenHelper GeneratorHelper may be a more appropriate name since it clearly says what is it about.

Fixed.

> Also it would be good to expand the imports such as this.

Fixed.

**Comment by indhub** [comment]
Installing JVM on CentOS:
```
sudo yum install java-1.7.0-openjdk
```

For windows, it is pretty much point click and install:
```
https://www.java.com/en/download/help/windows_manual_download.xml
```

I’m not sure it is easier to write the translator is Python because the target language is Python. 

That said, I don’t strongly lean one way or other in the choice between Java/Python. I preferred ANTLR/StringTemplates/Java simply because there was a lot of documentation and sample code and it was easy to get started. We could certainly spend some time writing a prototype in Python and evaluate if it makes sense to move to Python. That won’t be a quick work to do for the first release though. We can do it in a subsequent release.

File: tools/caffe_translator/README.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
+# Caffe Translator
+Caffe Translator is a source code translator that translates Caffe code into MXNet Python code. Note that this is different from the Caffe to MXNet model converted which is available [here](https://github.com/apache/incubator-mxnet/tree/master/tools/caffe_converter).
```
Comment (aaronmarkham) [inline tools/caffe_translator/README.md]: Consider naming it: "Caffe Code Translator" since a Caffe Translator exists in Caffe2 and it converts Caffe models to Caffe2 and doesn't have similar functionality to what is being presented here. I had to read this doc several times for it to sink in that these things were very different indeed. It's like a menu item called milkshake, but instead of ice cream, you get to drink some milk and someone grabs your head from behind and shakes it vigorously. If that's in the small text you might be surprised at the outcome.

File: tools/caffe_translator/README.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+```
+
+Here is the list of command line parameters accepted by the Caffe Translator:
+- *training-prototxt*: specifies the path to the training/validation prototxt to be translated.
+- *solver-prototxt*: specifies the path to the solver prototxt to be translated.
+- *output-file*: specifies the file to write the translated output into.
+- *params-file* (optional): specifies the .caffemodel file to initialize parameters from.
+- *custom-data-layers* (optional): Specifies a comma-separated list of types of the custom data layers used in the prototxt. The translator will use [`CaffeDataIter`](https://mxnet.incubator.apache.org/how_to/caffe.html#use-io-caffedataiter) to translate these layers to MXNet.
+
+**Note:** Translated code uses [`CaffeDataIter`](https://mxnet.incubator.apache.org/how_to/caffe.html#use-io-caffedataiter) to read from LMDB files. `CaffeDataIter` requires the number of examples in LMDB file to be specified as a parameter. You can provide this information before translation using a `#caffe2mxnet` directive like shown below:
```
Comment (aaronmarkham) [inline tools/caffe_translator/README.md]: Can you link to an asset that has mnist in a lmdb format for you? There's this one: http://download.caffe2.ai/databases/mnist-lmdb.zip
Or, if we don't want to link over there, why not load this file up on our own S3?

File: tools/caffe_translator/README.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+#### Run the translated code:
+
+Following prerequisites are required to run the translated code:
+1. Caffe with MXNet interface ([Why?](faq.md#why_caffe) [How to build?](https://github.com/apache/incubator-mxnet/tree/master/plugin/caffe#install-caffe-with-mxnet-interface))
+2. MXNet with Caffe plugin ([How to build?](https://github.com/apache/incubator-mxnet/tree/master/plugin/caffe#compile-with-caffe))
+
+Once prerequisites are installed, the translated Python code can be run like any other Python code:
+
+Example:
+```
```
Comment (aaronmarkham) [inline tools/caffe_translator/README.md]: I'm left wondering what the translated code looks like... what's the before and after? Can you put in examples here, and/or link to full files that I can review?

File: tools/caffe_translator/README.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
+# Caffe Translator
+Caffe Translator is a source code translator that translates Caffe code into MXNet Python code. Note that this is different from the Caffe to MXNet model converted which is available [here](https://github.com/apache/incubator-mxnet/tree/master/tools/caffe_converter).
+
```
Comment (aaronmarkham) [inline tools/caffe_translator/README.md]: You mention the "what" then move to the "how", but I think you need a "why". What's the use case here. I'm still not sure why you would do this.

File: tools/caffe_translator/README.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    backend: LMDB
+  }
+```
+
+#### Run the translated code:
+
+Following prerequisites are required to run the translated code:
+1. Caffe with MXNet interface ([Why?](faq.md#why_caffe) [How to build?](https://github.com/apache/incubator-mxnet/tree/master/plugin/caffe#install-caffe-with-mxnet-interface))
+2. MXNet with Caffe plugin ([How to build?](https://github.com/apache/incubator-mxnet/tree/master/plugin/caffe#compile-with-caffe))
+
```
Comment (aaronmarkham) [inline tools/caffe_translator/README.md]: Would (3) be datasets always in LMDB format? Or am I misunderstanding?

File: tools/caffe_translator/README.md (hunk 4 above)
Comment (indhub) [reply tools/caffe_translator/README.md]: Will it help if I change this to:

"Caffe Translator is a migration tool that helps developers migrate their existing Caffe code to MXNet and continue further development using MXNet. Note that this is different from the Caffe to MXNet model converter which is available [here](https://github.com/apache/incubator-mxnet/tree/master/tools/caffe_converter)."

I think mentioning that it is a migration tool to migrate code covers the "Why would you use this tool?" question.

File: tools/caffe_translator/README.md (hunk 3 above)
Comment (indhub) [reply tools/caffe_translator/README.md]: Here is an example:

Input:
https://github.com/BVLC/caffe/blob/master/examples/mnist/lenet_train_test.prototxt
https://github.com/BVLC/caffe/blob/master/examples/mnist/lenet_solver.prototxt

Output:
https://gist.github.com/indhub/85d8b82728f986a809b8cc99ab440792

I'll try to add this to the repo too. Only problem is I'll have to add apache license to the output if I have to put it in the repo and that could confuse users (make them wonder if Apache license applies to the generated code). Any suggestion to avoid that confusion (apart from explaining it in readme)?

File: tools/caffe_translator/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (indhub) [reply tools/caffe_translator/README.md]: Fixed.

Comment (indhub) [reply tools/caffe_translator/README.md]: Thanks for this feedback, Aaron. I’ve made changes in the second paragraph to include several samples of both input and output so that it is clear what the tool does. 

A name change will be hard to do this close to the release. But we can definitely discuss this post-release and make a name change if necessary.

File: tools/caffe_translator/README.md (hunk 4 above)
Comment (indhub) [reply tools/caffe_translator/README.md]: Made the change suggested above.

File: tools/caffe_translator/README.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
+# Caffe Translator
+Caffe Translator is a source code translator that translates Caffe code into MXNet Python code. Note that this is different from the Caffe to MXNet model converted which is available [here](https://github.com/apache/incubator-mxnet/tree/master/tools/caffe_converter).
+
+Caffe Translator takes the training/validation prototxt ([example](https://github.com/BVLC/caffe/blob/master/examples/mnist/lenet_train_test.prototxt)) and solver prototxt ([example](https://github.com/BVLC/caffe/blob/master/examples/mnist/lenet_solver.prototxt)) as input and produces MXNet Python code as output. The translated Python code uses MXNet Symbol and Module API to build the network, reads data from LMDB files, runs training and saves the trained model using the MXNet Module API.
```
[the remark this answers is not in the document]
Comment (indhub) [reply tools/caffe_translator/README.md]: Handled this in faq.

File: tools/caffe_translator/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (indhub) [reply tools/caffe_translator/README.md]: Changed to #CaffeToMXNet

Comment (indhub) [reply tools/caffe_translator/README.md]: Linked in the second paragraph.

File: tools/caffe_translator/README.md (hunk 3 above)
Comment (indhub) [reply tools/caffe_translator/README.md]: Added a link to the translated output in the second paragraph.

File: tools/caffe_translator/README.md [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+Once prerequisites are installed, the translated Python code can be run like any other Python code:
+
+Example:
+```
+python translated_code.py
+```
+
+### What layers are supported?
+
+Caffe Translator can currently translate the following layers.
```
[the remark this answers is not in the document]
Comment (indhub) [reply tools/caffe_translator/README.md]: Fixed

File: tools/caffe_translator/README.md [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+- Inner Product (Fully Connected layer)
+- Flatten
+- Permute
+- Pooling
+- Power
+- Relu
+- Scale<sup>*</sup>
+- SoftmaxOutput
+
+<sup>*</sup> - Uses [CaffePlugin](https://github.com/apache/incubator-mxnet/tree/master/plugin/caffe)
```
[the remark this answers is not in the document]
Comment (indhub) [reply tools/caffe_translator/README.md]: Fixed
